File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/javascript/reverse_string.js
function reverseString(str) {
  const provided_string = str
  const starting_array = str.split("")
  const ending_array = []

  if (provided_string === "") {
    return ""
  } else {
    while (starting_array.length > 0) {
      let last_char = starting_array[starting_array.length -1]
      ending_array.push(last_char)
      starting_array.pop()
    }
    return ending_array.join("")
  }
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'ih'");
  console.log("=>", reverseString("hi"));

  console.log("");

  console.log("Expecting: 'ybabtac'");
  console.log("=>", reverseString("catbaby"));

  console.log("");

  console.log("Expecting: 'tac-eiknarf'");
  console.log("=>", reverseString("frankie-cat"));

  console.log("");

  console.log("Expecting: '0987-654-321'");
  console.log("=>", reverseString("123-456-7890"));
}

module.exports = reverseString;

// Please add your working pseudocode to this file

// receive String in function invocation
// const string = "string"

// const array2 = string.split('')
// RETURNS ['s', 't', 'r', 'i', 'n', 'g']

// likely want to split string up so that each character is a value pushed in to the empty array
// declare empty final array
//   isolate last index of startting array
//   push into final array
//   AND pop off of starting array

//   loop through that until the starting array is empty
// return final array



// And a written explanation of your solution

// declare string variable with string provided in fxn invocation
// declare new array and split string into it
// declare final array that we'll push split string into
// if string is only one character, just return that character
// otherwise, while we still have letters left to reverse, we'll find the final character of the string, we'll push it into the final array, and we'll remove it from the starting array
// then after there are no more characters to reverse, we'll return the array BUT first we'll turn it back into a string by joining the value at each index

File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/ruby/reverse_string.rb
def reverse_string(str)
  # type your code in here
  reversed_string = ""

  str.chars.each do |char|
    reversed_string = char + reversed_string
  end

  reversed_string
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'ih'"
  puts "=>", reverse_string('hi')

  puts

  puts "Expecting: 'ybabtac'"
  puts "=>", reverse_string('catbaby')

  puts

  puts "Expecting: 'ybba'"
  puts "=>", reverse_string('abby')

  puts

  puts "Expecting: 'secnarf'"
  puts "=>", reverse_string('frances')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# set final reversed string variable equal to empty string
# then, for each character in the str from function invocation, do something
# with something being -- putting the character in front of whatever's already inside of the reversed string 
# then after it's done, return the reversed string

# And a written explanation of your solution
# Insted of converting to array and then back to string, we're declaring an empty string variable, and then iterating over each character of the given string. In the iteration, we'll add each character to the front of the new reversed string.
# So in the javascript method, we started with the last character and added them in reverse order onto the final product.
# In this method, we're starting with the first char and reordering them while adding to the final prorduct.